[Filters] Control applying the Filter by a new enum named FilterRenderingMode

shallawa opened the pull request:
#### ccb8d41a1adb174e09b49b63d2e91b46d2769e21
<pre>
[Filters] Control applying the Filter by a new enum named FilterRenderingMode
<a href="https://bugs.webkit.org/show_bug.cgi?id=248193">https://bugs.webkit.org/show_bug.cgi?id=248193</a>
rdar://102590236

Reviewed by Cameron McCormack.

RenderingMode has been used to control whether the Filter should use accelerated
or unaccelerated ImageBuffers. To support extended CoreGraphics filters, we need
even a new way to apply the filter which does not require creating temporary
ImageBuffers.

FilterRenderingMode will be used for this purpose. FilterEffect and Filter will
now provide OptionSet&lt;FilterRenderingMode&gt; telling which modes they support.
The Settings will provide preferred OptionSet&lt;FilterRenderingMode&gt;. Filter will
decide which FilterRenderingMode to use based on the preferred and the supported
FilterRenderingModes.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
(WebCore::Page::acceleratedFiltersEnabled const): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::supportedFilterRenderingModes const):
(WebCore::FEComponentTransfer::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::Filter):
(WebCore::Filter::renderingMode const):
(WebCore::Filter::setFilterRenderingMode):
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::filterRenderingMode const):
(WebCore::Filter::setFilterRenderingMode):
(WebCore::Filter::Filter):
(WebCore::Filter::renderingMode const): Deleted.
(WebCore::Filter::setRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::createApplier const):
(WebCore::FilterEffect::apply):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
(WebCore::FilterFunction::supportedFilterRenderingModes const):
(WebCore::FilterFunction::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterRenderingMode.h: Added.
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterRenderingModes const):
(WebCore::SourceGraphic::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/SourceGraphic.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
(WebCore::CSSFilter::CSSFilter):
(WebCore::createReferenceFilter):
(WebCore::CSSFilter::buildFilterFunctions):
(WebCore::CSSFilter::supportedFilterRenderingModes const):
(WebCore::CSSFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::image const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
(WebCore::SVGFilter::supportedFilterRenderingModes const):
(WebCore::SVGFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/Platform/IPC/FilterReference.h:
(IPC::FilterReference::encodeFilter):
(IPC::FilterReference::decodeFilter):

Canonical link: <a href="https://commits.webkit.org/256985@main">https://commits.webkit.org/256985@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fd12326e2c7bfa1349d7d5c1ef91d3ec393fce2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6747 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30654 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106997 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/167258 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7078 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35500 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89888 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103668 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103148 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84121 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32309 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88970 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75235 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/752 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/738 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21910 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4816 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5549 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1989 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41277 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/filters/Filter.h:
```diff
@@ -53,6 +53,8 @@ class Filter : public FilterFunction {
     ClipOperation clipOperation() const { return m_clipOperation; }
     void setClipOperation(ClipOperation clipOperation) { m_clipOperation = clipOperation; }
 
+    RenderingMode renderingMode() const { return m_filterMode == FilterMode::Accelerated ? RenderingMode::Accelerated : RenderingMode::Unaccelerated; }
```
Is this only used to determine the type of ImageBuffer to create? Can we use a name that indicates that, like `imageBufferRenderingMode`? Also, should we be never calling this when `m_filterMode` is `FilterMode::GraphicsContext`, and if so, should we assert that in here?

shallawa replied:
I renamed `FilterMode` to `FilterRenderingMode`. So I think the distinction between `FilterRenderingMode` and `RenderingMode` should be obvious.

Yes an assert will be added in `m_filterRenderingMode` is `FilterRenderingMode::GraphicsContext.`

heycam commented on Source/WebCore/rendering/CSSFilter.cpp:
```diff
@@ -42,21 +42,21 @@
 
-    if (renderingMode == RenderingMode::Accelerated && !filter->supportsAcceleratedRendering())
-        filter->setRenderingMode(RenderingMode::Unaccelerated);
+    if (preferredFilterMode != FilterMode::Software && !filter->supportedFilterModes().contains(preferredFilterMode))
```
Can drop the `preferredFilterMode != FilterMode::Software` part of this, since `supportedFilterModes()` should always include it.

shallawa replied:
I removed it although I added it as an optimization.

heycam commented on Source/WebCore/rendering/CSSFilter.cpp:
```diff
@@ -315,17 +315,19 @@ FilterEffectVector CSSFilter::effectsOfType(FilterFunction::Type filterType) con
     for (auto& function : m_functions) {
-        if (!function->supportsAcceleratedRendering())
-            return false;
+        auto functionModes = function->supportedFilterModes();
```
Should we assert in here that `functionModes` is not empty?

heycam commented on Source/WebCore/rendering/CSSFilter.cpp:
```diff
@@ -315,17 +315,19 @@ FilterEffectVector CSSFilter::effectsOfType(FilterFunction::Type filterType) con
+        auto functionModes = function->supportedFilterModes();
+        if (!filterModes)
+            filterModes = functionModes;
+        else
```
Instead of this, we could start with `filterModes` with all three bits set, then we wouldn't need to handle the first function specially.

shallawa replied:
This is done by initializing filterModes with AllFilterRenderingModes which is defined to be
```
constexpr OptionSet<FilterRenderingMode> AllFilterRenderingModes = {
    FilterRenderingMode::Software,
    FilterRenderingMode::Accelerated,
    FilterRenderingMode::GraphicsContext
};
```

heycam commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -208,18 +208,19 @@ FloatPoint3D SVGFilter::resolvedPoint3D(const FloatPoint3D& point) const
-            return false;
+        auto effectModes = term.effect->supportedFilterModes();
+        if (!filterModes)
+            filterModes = effectModes;
```
Same comments as on `CSSFilter::supportedFilterModes`.

heycam commented on Source/WebCore/platform/graphics/filters/FilterMode.h:
```diff
@@ -0,0 +1,50 @@
+    Accelerated     = 1 << 0,
+    GraphicsContext = 1 << 1,
+    Software        = 1 << 2
+};
```
Maybe `FilterRenderingMode`? To me, `FilterMode` is a bit generic, and it could be referring to anything related to filter handling.

shallawa replied:
Done.

heycam commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -47,8 +47,8 @@ RefPtr<SVGFilter> SVGFilter::create(SVGFilterElement& filterElement, RenderingMo
 
-    if (renderingMode == RenderingMode::Accelerated && !filter->supportsAcceleratedRendering())
-        filter->setRenderingMode(RenderingMode::Unaccelerated);
+    if (preferredFilterMode != FilterMode::Software && !filter->supportedFilterModes().contains(preferredFilterMode))
```
Same comment as on `CSSFilter::create`.

heycam commented on Source/WebCore/platform/graphics/filters/FilterFunction.h:
```diff
@@ -93,7 +94,7 @@ class FilterFunction : public RefCounted<FilterFunction> {
     AtomString filterName() const { return filterName(m_filterType); }
 
-    virtual bool supportsAcceleratedRendering() const { return false; }
+    virtual OptionSet<FilterRenderingMode> supportedFilterRenderingModes() const { return { FilterRenderingMode::Software }; }
```
Nit: I think the braces around `FilterRenderingMode::Software` aren't needed, since `OptionSet` has an implicit constructor that takes the enum type.

heycam commented on Source/WebCore/platform/graphics/filters/FilterRenderingMode.h:
```diff
@@ -0,0 +1,58 @@
+    GraphicsContext = 1 << 2
+};
+
+constexpr OptionSet<FilterRenderingMode> AllFilterRenderingModes = {
```
Nit: as a variable, this should start with a lowercase letter.

shallawa committed:
[Filters] Control applying the Filter by a new enum named FilterRenderingMode
https://bugs.webkit.org/show_bug.cgi?id=248193
rdar://102590236

Reviewed by Cameron McCormack.

RenderingMode has been used to control whether the Filter should use accelerated
or unaccelerated ImageBuffers. To support extended CoreGraphics filters, we need
even a new way to apply the filter which does not require creating temporary
ImageBuffers.

FilterRenderingMode will be used for this purpose. FilterEffect and Filter will
now provide OptionSet<FilterRenderingMode> telling which modes they support.
The Settings will provide preferred OptionSet<FilterRenderingMode>. Filter will
decide which FilterRenderingMode to use based on the preferred and the supported
FilterRenderingModes.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
(WebCore::Page::acceleratedFiltersEnabled const): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/graphics/filters/FEColorMatrix.cpp:
(WebCore::FEColorMatrix::supportedFilterRenderingModes const):
(WebCore::FEColorMatrix::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEColorMatrix.h:
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::supportedFilterRenderingModes const):
(WebCore::FEComponentTransfer::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
* Source/WebCore/platform/graphics/filters/Filter.cpp:
(WebCore::Filter::Filter):
(WebCore::Filter::renderingMode const):
(WebCore::Filter::setFilterRenderingMode):
* Source/WebCore/platform/graphics/filters/Filter.h:
(WebCore::Filter::filterRenderingMode const):
(WebCore::Filter::setFilterRenderingMode):
(WebCore::Filter::Filter):
(WebCore::Filter::renderingMode const): Deleted.
(WebCore::Filter::setRenderingMode): Deleted.
* Source/WebCore/platform/graphics/filters/FilterEffect.cpp:
(WebCore::FilterEffect::createApplier const):
(WebCore::FilterEffect::apply):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
(WebCore::FilterFunction::supportedFilterRenderingModes const):
(WebCore::FilterFunction::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterRenderingMode.h: Added.
* Source/WebCore/platform/graphics/filters/SourceGraphic.cpp:
(WebCore::SourceGraphic::supportedFilterRenderingModes const):
(WebCore::SourceGraphic::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/platform/graphics/filters/SourceGraphic.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::CSSFilter::create):
(WebCore::CSSFilter::CSSFilter):
(WebCore::createReferenceFilter):
(WebCore::CSSFilter::buildFilterFunctions):
(WebCore::CSSFilter::supportedFilterRenderingModes const):
(WebCore::CSSFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/rendering/CSSFilter.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerFilters.cpp:
(WebCore::RenderLayerFilters::beginFilterEffect):
* Source/WebCore/rendering/RenderLayerFilters.h:
* Source/WebCore/rendering/style/StyleFilterImage.cpp:
(WebCore::StyleFilterImage::image const):
* Source/WebCore/rendering/svg/RenderSVGResourceFilter.cpp:
(WebCore::RenderSVGResourceFilter::applyResource):
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
(WebCore::writeSVGResourceContainer):
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::create):
(WebCore::SVGFilter::SVGFilter):
(WebCore::SVGFilter::supportedFilterRenderingModes const):
(WebCore::SVGFilter::supportsAcceleratedRendering const): Deleted.
* Source/WebCore/svg/graphics/filters/SVGFilter.h:
* Source/WebKit/Platform/IPC/FilterReference.h:
(IPC::FilterReference::encodeFilter):
(IPC::FilterReference::decodeFilter):

Canonical link: https://commits.webkit.org/256985@main
